README.md

```diff
@@ -0,0 +1,834 @@
+
+export class A11yInput extends LitElement {}
+
+customElements.define('a11y-input', A11yInput);
+```
+
+and `test/a11y-input.test.js`;
+```js
+/* eslint-disable no-unused-expressions */
+import { html, fixture, expect } from '@open-wc/testing';
+
+import '../src/a11y-input.js';
+
+/**
+ * @typedef {import('../src/a11y-input.js').A11yInput} A11yInput
+ */
+
+describe('a11y input', () => {
+  it('has by default an empty string as label', async () => {
+    const el = /** @type {A11yInput} */ (await fixture('<a11y-input></a11y-input>'));
+    expect(el.label).to.equal('');
+  });
+});
+```
+
+Let's see if everything works correctly by running: `npm run test`.
+
+```
+SUMMARY:
+✔ 0 tests completed
+✖ 1 test failed
+
+FAILED TESTS:
+  a11y input
+    ✖ has by default an empty string as label
+      HeadlessChrome 73.0.3683 (Windows 10.0.0)
+    AssertionError: expected undefined to equal ''
+
+      + expected - actual
+
+      -[undefined]
+      +""
+```
+
+Awesome - just as expected, we have a failing test :)
+
+Let's switch into watch mode `npm run test:watch`, this will run the tests simultaneously when you make some changes to your code.
+
+![01-watch-mode-intro](https://github.com/daKmoR/testing-workflow-for-web-components/raw/master/images/01-watch-mode-intro.gif)
+
+So far so good? Still with us? Great! Let's up the game a little.
+
+### Adding a test for shadow dom
+
+```js
+it('has a static shadowDom', async () => {
+  const el = /** @type {A11yInput} */ (await fixture(html`
+    <a11y-input></a11y-input>
+  `));
+  expect(el.shadowRoot.innerHTML).to.equal(`
+    <slot name="label"></slot>
+    <slot name="input"></slot>
+  `);
+});
+```
+
+As expected we get
+```
+✖ has a static shadowDom
+AssertionError: expected '' to equal '\n      <slot name="label"></slot>\n      <slot name="input"></slot>\n    '
+
+  + expected - actual
+
+  +
+  +      <slot name="label"></slot>
+  +      <slot name="input"></slot>
+  +
+```
+
+Add the implementation
+```js
+render() {
+  return html`
+    <slot name="label"></slot>
+    <slot name="input"></slot>
+  `;
+}
+```
+
+Interesting, the test should be green... but it's not :/ Let's take a look.
+```
+✖ has a static shadowDom
+AssertionError: expected '<!---->\n      <slot name="label"></slot>\n      <slot name="input"></slot>\n    <!---->' to equal '\n        <slot name="label"></slot>\n        <slot name="input"></slot>\n    '
+
+  + expected - actual
+
+  -<!---->
+  -      <slot name="label"></slot>
+  -      <slot name="input"></slot>
+  -    <!---->
+  +
+  +        <slot name="label"></slot>
+  +        <slot name="input"></slot>
+  +
+```
+
+You may have noticed those weird empty comment `<!---->` tags. They are markers that `lit-html` uses to remember where dynamic parts are, so it can be updated efficiently. For testing however, this can be a little annoying to deal with.
+
+Do we really need to match the exact same indentation in code and test, and incorporate those markers?
```

Comment:
```suggestion
If we use `innerHTML` to compare the DOM, we'd have to rely on simple string equality. Under those circumstances, we'd have to exactly match the generated DOM's whitespace, comments, etc; in other words: it will need to be a perfect match. Really all we need to test is that the elements we want to render are rendered. We want to test the *semantic* contents of the shadow root.
```